IPython/core/completer.py

```diff
@@ -417,6 +537,189 @@ def __hash__(self):
         return hash((self.start, self.end, self.text))
 
 
+class SimpleCompletion:
+    """Completion item to be included in the dictionary returned by new-style Matcher (API v2).
+
+    .. warning::
+
+        Provisional
+
+        This class is used to describe the currently supported attributes of
+        simple completion items, and any additional implementation details
+        should not be relied on. Additional attributes may be included in
+        future versions, and meaning of text disambiguated from the current
+        dual meaning of "text to insert" and "text to used as a label".
+    """
+
+    __slots__ = ["text", "type"]
+
+    def __init__(self, text: str, *, type: str = None):
+        self.text = text
+        self.type = type
+
+    def __repr__(self):
+        return f"<SimpleCompletion text={self.text!r} type={self.type!r}>"
+
+
+class _MatcherResultBase(TypedDict):
+    """Definition of dictionary to be returned by new-style Matcher (API v2)."""
+
+    #: Suffix of the provided ``CompletionContext.token``, if not given defaults to full token.
+    matched_fragment: NotRequired[str]
+
+    #: Whether to suppress results from all other matchers (True), some
+    #: matchers (set of identifiers) or none (False); default is False.
+    suppress: NotRequired[Union[bool, Set[str]]]
+
+    #: Identifiers of matchers which should NOT be suppressed when this matcher
+    #: requests to suppress all other matchers; defaults to an empty set.
+    do_not_suppress: NotRequired[Set[str]]
+
+    #: Are completions already ordered and should be left as-is? default is False.
+    ordered: NotRequired[bool]
+
+
+@sphinx_options(show_inherited_members=True, exclude_inherited_from=["dict"])
+class SimpleMatcherResult(_MatcherResultBase, TypedDict):
```

Comment:
DO you need to inherit both if `_MatcherResultBase` is already TypedDict. 
and do we want to TypedDict, or simply make a datalass with all the possible existing fields, even if it's slightly incompatible ?

Reply:
> DO you need to inherit both if _MatcherResultBase is already TypedDict.

No, not at runtime, but yes if we want to have nice documentation (https://github.com/sphinx-doc/sphinx/pull/10806); a quirk of how `TypedDict` it was implemented. I added a comment in the body on this.

> and do we want to TypedDict, or simply make a datalass with all the possible existing fields, even if it's slightly backward-incompatible ?

I hesitated here and considered making it a dataclass to, but I preferred `TypedDict` for two reasons:
- in future releases we can add a new `NotRequired` field without breaking compatibility (and give notice if we want to make it required in the further future)
- it does not require users to import any specific class from IPython to wrap the results (or create a custom class complying with the API), so it would keep it simple (while argument name and type safety is provided by typecheckers)